File: HW/HW7/HW7.py
import heapq

class Main:
    def minMeetingRooms(self, intervals):
        if not intervals:
            return 0
        
        intervals.sort(key=lambda x: x[0])
        end_times = []
        heapq.heappush(end_times, intervals[0][1])
        
        for i in range(1, len(intervals)):
            start, end = intervals[i]
            earliest_end_time = end_times[0]
            
            if earliest_end_time <= start:
                heapq.heappop(end_times)
            heapq.heappush(end_times, end)
        
        return len(end_times)

# Test the solution
if __name__ == "__main__":
    main = Main()
    intervals = [[0, 30], [5, 10], [15, 20]]
    print(Main.minMeetingRooms(intervals)) 

File: HW/HW7/HW7.java
import java.util.Arrays;
import java.util.PriorityQueue;
public class main {
public int minMeetingRooms(int[][] intervals) {  // Your code here 
    if (intervals == null || intervals.length == 0) {
        return 0;
    }
    
    Arrays.sort(intervals, (a, b) -> a[0] - b[0]);

    PriorityQueue<Integer> endTimes = new PriorityQueue<>();
    endTimes.offer(intervals[0][1]);
    
    // Iterate through the rest of the meetings
    for (int i = 1; i < intervals.length; i++) {
        int[] currentJob = intervals[i];
        int earliestEndTime = endTimes.peek();
        if (earliestEndTime <= currentJob[0]) {
            endTimes.poll();
        }
        

        endTimes.offer(currentJob[1]);
    }
    
    return endTimes.size();



}
}
